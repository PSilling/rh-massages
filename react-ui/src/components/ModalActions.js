// react imports
import React, { Component } from 'react';

// util imports
import _t from '../utils/Translations.js';

/**
 * Custom component rendering action buttons for modal footers.
 */
class ModalActions extends Component {

  render() {
    return(
      <div className="pull-right">
        <button type="button" className="btn btn-primary" onClick={this.props.onProceed}
          style={{ 'marginRight': '5px' }} autoFocus={this.props.autoFocus}>
          {this.props.primaryLabel}
        </button>
        <button type="button" className="btn btn-default"
          onClick={this.props.onClose}>
          { _t.translate('Dismiss') }
        </button>
      </div>
    );
  }
}

export default ModalActions
