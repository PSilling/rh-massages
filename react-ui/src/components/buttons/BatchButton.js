// react imports
import React, { Component } from 'react';
import PropTypes from 'prop-types';

/**
 * A default styled button component with a given label. Used primarily for
 * batch operation buttons.
 */
class BatchButton extends Component {

  render() {
    return (
      <span>
        <button type="button" className="btn btn-default" onClick={this.props.onClick}
          disabled={this.props.disabled}>
          {this.props.label}
        </button>
      </span>
    )
  }
}

BatchButton.propTypes = {
  /** function to be called on button click */
  onClick: PropTypes.func.isRequired,
   /** button label */
  label: PropTypes.string,
  /** whether the button should be disabled */
  disabled: PropTypes.bool
};

BatchButton.defaultProps = {
  disabled: false
};

export default BatchButton
