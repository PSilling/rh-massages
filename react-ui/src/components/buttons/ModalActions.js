// react imports
import React from "react";
import PropTypes from "prop-types";

// util imports
import _t from "../../util/Translations";

/**
 * Footer buttons for Modal dialogs. Children components can be given to add more
 * case specific elements.
 */
const ModalActions = function ModalActions(props) {
  return (
    <div className="pull-right">
      {props.children}
      {props.primaryLabel === "none" ? (
        ""
      ) : (
        <button
          type="button"
          className="btn btn-primary"
          onClick={props.onProceed}
          title={props.title}
          style={{ marginRight: "5px" }}
          autoFocus={props.autoFocus}
          disabled={props.disabled}
        >
          {props.primaryLabel}
        </button>
      )}
      <button type="button" className="btn btn-default" onClick={props.onClose}>
        {_t.translate("Dismiss")}
      </button>
    </div>
  );
};

ModalActions.propTypes = {
  /** callback function triggered on close button click */
  onClose: PropTypes.func.isRequired,
  /** callback function triggered on primary button click */
  onProceed: PropTypes.func.isRequired,
  /** whether the primary button should be automatically focused */
  autoFocus: PropTypes.bool,
  /** extra elements to be provided before the two buttons */
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]),
  /** whether the primary button should be disabled */
  disabled: PropTypes.bool,
  /** primary button label to be displayed */
  primaryLabel: PropTypes.string,
  /** primary button title */
  title: PropTypes.string
};

ModalActions.defaultProps = {
  autoFocus: false,
  children: null,
  disabled: false,
  primaryLabel: _t.translate("Proceed"),
  title: ""
};

export default ModalActions;
