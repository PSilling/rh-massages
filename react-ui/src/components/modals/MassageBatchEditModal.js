// react imports
import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';

// component imports
import BatchButton from '../buttons/BatchButton';
import ModalActions from '../buttons/ModalActions';

// module imports
import Datetime from 'react-datetime';
import {ModalContainer, ModalDialog} from 'react-modal-dialog';
import moment from 'moment';

// util imports
import _t from '../../util/Translations';
import Util from '../../util/Util';

/**
 * Input Modal for editing multiple rules of Massages at once.
 */
class MassageBatchEditModal extends Component {

  state = {editMasseuse: false, editDate: 1, removeClients: false,
            masseuse: "", days: "1", time: moment("00:00", "HH:mm")}

   componentWillReceiveProps(nextProps) {
     if (this.props === nextProps) return;
     this.setState({
       masseuse: Util.isEmpty(nextProps.massages[0]) ? "" : nextProps.massages[0].masseuse,
       removeClients: false
     });
   }

  changeEditMasseuse = (event) => {
    this.setState({editMasseuse: event.target.checked});
  }

  changeEditDate = (event, dateCheckbox) => {
    if (dateCheckbox) {
      if (event.target.checked) {
        this.setState({editDate: 1});
      } else {
        this.setState({editDate: 0});
      }
    } else if (event.target.checked) {
      this.setState({editDate: event.target.value});
    }
  }

  changeRemoveClients = (event) => {
    this.setState({removeClients: event.target.checked});
  }

  changeMasseuse = (event) => {
    this.setState({masseuse: event.target.value});
  }

  changeDays = (event) => {
    if (Util.isEmpty(event.target.value) || parseInt(event.target.value, 10) < 0
        || parseInt(event.target.value, 10) > 365) {
      return;
    }
    this.setState({days: event.target.value});
  }

  changeTime = (time) => {
    if (typeof time === 'string') {
      return;
    }
    this.setState({time: time});
  }

  /**
   * Get date value based on a given offset and state time.
   */
  getDate = (date) => {
    var minutes = (parseInt(this.state.days, 10) * 1440) +
      (this.state.time.get('hour') * 60) + this.state.time.get('minute');
    if (this.state.editDate < 0) {
      if (moment(date).subtract(minutes, 'minutes').isBefore(moment().startOf('minute'))) {
        return -1;
      }
      return moment(date).subtract(minutes, 'minutes').toDate();
    } else if (this.state.editDate > 0) {
      return moment(date).add(minutes, 'minutes').toDate();
    } else {
      return date;
    }
  }

  /**
   * Edits all Massages based on supplied state values.
   */
  editMassages = () => {
    var putArray = [],
        informed = false;
    for (var i = 0; i < this.props.massages.length; i++) {
      if (this.getDate(this.props.massages[i].date) === -1) {
        if (!informed) {
          Util.notify("warning",
            _t.translate('Not all massages were edited as in some cases the new date would have been before now.'),
            _t.translate('Warning'));
          informed = true;
        }
        continue;
      }
      putArray.push({
        id: this.props.massages[i].id,
        date: this.getDate(this.props.massages[i].date),
        ending: this.getDate(this.props.massages[i].ending),
        masseuse: (this.state.editMasseuse) ? this.state.masseuse : this.props.massages[i].masseuse,
        client: (this.state.removeClients) ? null : this.props.massages[i].client,
        facility: {id: this.props.massages[i].facility.id}
      });
    }
    if (putArray.length > 0) {
      Util.put(Util.MASSAGES_URL, putArray, () => {
        this.props.getCallback();
        this.props.onToggle(true);
      });
    }
  }

  handleModalKeyPress = (event) => {
    if (event.charCode === 13 && document.activeElement === ReactDOM.findDOMNode(this.modalDialog)) {
      this.editMassages();
    }
  }

  handleInputKeyPress = (event) => {
    if (event.key === 'Enter') {
      this.editMassages();
    }
  }

  render() {
    return (
      <span style={{ 'marginRight': '5px' }}>
        <BatchButton onClick={() => this.props.onToggle(false)} disabled={this.props.disabled}
          label={ _t.translate('Edit selected') } />

        {this.props.active ?
          <ModalContainer onClose={() => this.props.onToggle(false)}>
            <ModalDialog onClose={() => this.props.onToggle(false)} width="50%" style={{ 'outline': 'none' }}
              tabIndex="1" onKeyPress={this.handleModalKeyPress}
              ref={(dialog) => {
                this.modalDialog = dialog;
              }}>
              <h2>
                { _t.translate('Edit Massages') }
              </h2>
              <hr />
              <div className="form-group col-md-12">
                <label className="checkbox-inline">
                  <input type="checkbox" onChange={this.changeEditMasseuse}
                    checked={this.state.editMasseuse} style={{ 'marginRight': '5px' }}
                    onKeyPress={this.handleInputKeyPress}
                  />
                  <strong>{ _t.translate('Change masseur/masseuse') }</strong>
                </label>
                <input value={this.state.masseuse} onChange={this.changeMasseuse}
                  className="form-control" onFocus={Util.moveCursorToEnd}
                  onKeyPress={this.handleInputKeyPress} type="text" maxLength="64"
                  placeholder={ _t.translate('Masseur/Masseuse') } list="masseuses"
                  disabled={!this.state.editMasseuse}
                />
                <datalist id="masseuses">
                  {this.props.masseuses.map((item) => (
                    <option key={item} value={item} />
                  ))}
                </datalist>
              </div>

              <div className="form-group col-md-12">
                <label className="checkbox-inline">
                  <input type="checkbox" onChange={(event) => this.changeEditDate(event, true)}
                    onKeyPress={this.handleInputKeyPress} style={{ 'marginRight': '5px' }}
                    checked={this.state.editDate === 0 ? false : true}
                  />
                  <strong>{ _t.translate('Shift massage date (day and time)') }</strong>
                </label>
                <div className="row">
                  <div className="col-md-2">
                    <input value={this.state.days} onChange={this.changeDays} className="form-control"
                      onKeyPress={this.handleInputKeyPress} autoFocus onFocus={Util.moveCursorToEnd}
                      type="number" min="0" max="365" placeholder={ _t.translate('Day change') }
                      title={ _t.translate('Days') }
                      disabled={this.state.editDate === 0 ? true : false}
                    />
                  </div>

                  <div className="col-md-3">
                    <Datetime value={this.state.time} onChange={this.changeTime} dateFormat={false}
                      inputProps={{
                        placeholder: _t.translate('Duration change'),
                        onKeyPress: this.handleInputKeyPress,
                        disabled: (this.state.editDate === 0 ? true : false)
                      }}
                    />
                  </div>

                  <div className="col-md-4" style={{ 'marginTop': '5px' }}>
                    <label className="radio-inline">
                      <input type="radio" onChange={(event) => this.changeEditDate(event, false)} value={1}
                        checked={this.state.editDate > 0 ? true : false} onKeyPress={this.handleInputKeyPress}
                        disabled={this.state.editDate === 0 ? true : false}
                        style={{ 'marginRight': '5px' }}
                      />
                      <strong>{ _t.translate('Later') }</strong>
                    </label>
                    <label className="radio-inline">
                      <input type="radio" onChange={(event) => this.changeEditDate(event, false)} value={-1}
                        checked={this.state.editDate < 0 ? true : false} onKeyPress={this.handleInputKeyPress}
                        disabled={this.state.editDate === 0 ? true : false}
                        style={{ 'marginRight': '5px' }}
                      />
                      <strong>{ _t.translate('Earlier') }</strong>
                    </label>
                  </div>
                </div>
              </div>

              <div className="form-group col-md-12">
                <label className="checkbox-inline">
                  <input type="checkbox" onChange={this.changeRemoveClients}
                    checked={this.state.removeClients} style={{ 'marginRight': '5px' }}
                    onKeyPress={this.handleInputKeyPress}
                  />
                  <strong>{ _t.translate('Also remove all assigned clients') }</strong>
                </label>
              </div>
              <ModalActions
                primaryLabel={ _t.translate('Edit') }
                onProceed={this.editMassages}
                onClose={() => this.props.onToggle(false)}
                autoFocus={false}
              />
            </ModalDialog>
          </ModalContainer> : ''
        }
      </span>
    )
  }
}

MassageBatchEditModal.propTypes = {
  /** whether the dialog should be shown */
  active: PropTypes.bool,
  /** whether the trigger button should be disabled or not */
  disabled: PropTypes.bool,
  /** Massages to be copied */
  massages: PropTypes.arrayOf(PropTypes.object).isRequired,
  /** unique Massage masseuses of the given Facility */
  masseuses: PropTypes.arrayOf(PropTypes.string),
  /** callback function for Massage list update */
  getCallback: PropTypes.func.isRequired,
  /** function called on modal toggle */
  onToggle: PropTypes.func.isRequired
};

export default MassageBatchEditModal
