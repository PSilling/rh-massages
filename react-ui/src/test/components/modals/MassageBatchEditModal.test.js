// react imports
import React from 'react';
import TestRenderer from 'react-test-renderer';

// test imports
import BatchButton from '../../../components/buttons/BatchButton';
import MassageBatchEditModal from '../../../components/modals/MassageBatchEditModal';
import ModalActions from '../../../components/buttons/ModalActions';

// test mocks
jest.mock('../../../util/Util');

afterEach(() => {
  jest.resetAllMocks();
});

test('renders inside content with correct props', () => {
  const testGetFunction = jest.fn(),
        testToggleFunction = jest.fn(),
        testMassages = [{
          id: 1,
          date: new Date(0),
          ending: new Date(1000),
          client: null,
          facility: { id: 1, name: "test" }
        }],
        testMasseuses = ["test"],
        testRenderer = TestRenderer.create(
          <MassageBatchEditModal active massages={testMassages} masseuses={testMasseuses}
            getCallback={testGetFunction} onToggle={testToggleFunction} withPortal={false} />
        ),
        testInstance = testRenderer.root;

  let button = testInstance.findByType(BatchButton),
      actions = testInstance.findByType(ModalActions),
      datalistOption = testInstance.findByType('option'),
      treeJSON = testRenderer.toJSON();

  expect(testToggleFunction).not.toHaveBeenCalled();

  button.props.onClick();
  actions.props.onClose();

  expect(testToggleFunction).toHaveBeenCalledTimes(2);
  expect(testToggleFunction).toHaveBeenLastCalledWith(false);
  expect(testGetFunction).not.toHaveBeenCalled();

  actions.props.onProceed();

  expect(testToggleFunction).toHaveBeenCalledTimes(3);
  expect(testToggleFunction).toHaveBeenLastCalledWith(true);
  expect(testGetFunction).toHaveBeenCalledTimes(1);
  expect(button.props.disabled).toBe(false);
  expect(datalistOption.props.value).toBe("test");
  expect(treeJSON).toMatchSnapshot();
});
