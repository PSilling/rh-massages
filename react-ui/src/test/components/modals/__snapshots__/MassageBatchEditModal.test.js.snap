// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders inside content with correct props 1`] = `
<span
  style={
    Object {
      "marginRight": "5px",
    }
  }
>
  <span>
    <button
      className="btn btn-default"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Upravit vybrané
    </button>
  </span>
  <div>
    <h2>
      Upravit masáže
    </h2>
    <hr />
    <div
      className="form-group col-md-12"
    >
      <label
        className="checkbox-inline"
      >
        <input
          checked={false}
          onChange={[Function]}
          onKeyPress={[Function]}
          style={
            Object {
              "marginRight": "5px",
            }
          }
          type="checkbox"
        />
        <strong>
          Změnit maséra/masérku
        </strong>
      </label>
      <input
        className="form-control"
        disabled={true}
        list="masseuses"
        maxLength="64"
        onChange={[Function]}
        onFocus={undefined}
        onKeyPress={[Function]}
        placeholder="Masér/Masérka"
        type="text"
        value=""
      />
      <datalist
        id="masseuses"
      >
        <option
          value="test"
        />
      </datalist>
    </div>
    <div
      className="form-group col-md-12"
    >
      <label
        className="checkbox-inline"
      >
        <input
          checked={true}
          onChange={[Function]}
          onKeyPress={[Function]}
          style={
            Object {
              "marginRight": "5px",
            }
          }
          type="checkbox"
        />
        <strong>
          Posunout datum masáže (den a čas)
        </strong>
      </label>
      <div
        className="row"
      >
        <div
          className="col-md-2"
        >
          <input
            autoFocus={true}
            className="form-control"
            disabled={false}
            max="365"
            min="0"
            onChange={[Function]}
            onFocus={undefined}
            onKeyPress={[Function]}
            placeholder="Změna dne"
            title="Dny"
            type="number"
            value="1"
          />
        </div>
        <div
          className="col-md-3"
        >
          <div
            className="rdt"
          >
            <input
              className="form-control"
              disabled={false}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyPress={[Function]}
              placeholder="Změna trvání"
              type="text"
              value="12:00 AM"
            />
            <div
              className="rdtPicker"
            >
              <div
                className="rdtTime"
              >
                <table>
                  <tbody>
                    <tr>
                      <td>
                        <div
                          className="rdtCounters"
                        >
                          <div
                            className="rdtCounter"
                          >
                            <span
                              className="rdtBtn"
                              onContextMenu={[Function]}
                              onMouseDown={[Function]}
                              onTouchStart={[Function]}
                            >
                              ▲
                            </span>
                            <div
                              className="rdtCount"
                            >
                              12
                            </div>
                            <span
                              className="rdtBtn"
                              onContextMenu={[Function]}
                              onMouseDown={[Function]}
                              onTouchStart={[Function]}
                            >
                              ▼
                            </span>
                          </div>
                          <div
                            className="rdtCounterSeparator"
                          >
                            :
                          </div>
                          <div
                            className="rdtCounter"
                          >
                            <span
                              className="rdtBtn"
                              onContextMenu={[Function]}
                              onMouseDown={[Function]}
                              onTouchStart={[Function]}
                            >
                              ▲
                            </span>
                            <div
                              className="rdtCount"
                            >
                              00
                            </div>
                            <span
                              className="rdtBtn"
                              onContextMenu={[Function]}
                              onMouseDown={[Function]}
                              onTouchStart={[Function]}
                            >
                              ▼
                            </span>
                          </div>
                          <div
                            className="rdtCounter"
                          >
                            <span
                              className="rdtBtn"
                              onContextMenu={[Function]}
                              onMouseDown={[Function]}
                              onTouchStart={[Function]}
                            >
                              ▲
                            </span>
                            <div
                              className="rdtCount"
                            >
                              AM
                            </div>
                            <span
                              className="rdtBtn"
                              onContextMenu={[Function]}
                              onMouseDown={[Function]}
                              onTouchStart={[Function]}
                            >
                              ▼
                            </span>
                          </div>
                        </div>
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>
          </div>
        </div>
        <div
          className="col-md-4"
          style={
            Object {
              "marginTop": "5px",
            }
          }
        >
          <label
            className="radio-inline"
          >
            <input
              checked={true}
              disabled={false}
              onChange={[Function]}
              onKeyPress={[Function]}
              style={
                Object {
                  "marginRight": "5px",
                }
              }
              type="radio"
              value={1}
            />
            <strong>
              Později
            </strong>
          </label>
          <label
            className="radio-inline"
          >
            <input
              checked={false}
              disabled={false}
              onChange={[Function]}
              onKeyPress={[Function]}
              style={
                Object {
                  "marginRight": "5px",
                }
              }
              type="radio"
              value={-1}
            />
            <strong>
              Dříve
            </strong>
          </label>
        </div>
      </div>
    </div>
    <div
      className="form-group col-md-12"
    >
      <label
        className="checkbox-inline"
      >
        <input
          checked={false}
          onChange={[Function]}
          onKeyPress={[Function]}
          style={
            Object {
              "marginRight": "5px",
            }
          }
          type="checkbox"
        />
        <strong>
          Také odstranit všechny zapsané klienty
        </strong>
      </label>
    </div>
    <div
      className="pull-right"
    >
      <button
        autoFocus={false}
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "marginRight": "5px",
          }
        }
        title=""
        type="button"
      >
        Upravit
      </button>
      <button
        className="btn btn-default"
        onClick={[Function]}
        type="button"
      >
        Zrušit
      </button>
    </div>
  </div>
</span>
`;
